### PR TITLE
Package Index: Build in Dockerhub

### PR DESCRIPTION
**Pull request by ax3l**
Prepare to build the package index service, packages.spack.io, on Dockerhub.

Local build (in spack root dir):
```
docker build -t spack/packages.spack.io:latest -f share/spack/docker/package-index/Dockerfile .
```

Local test:
```
docker run -p 8080:80 spack/packages.spack.io:latest
```

- [x] Removes leftover docker build stages from Travis CI.
- [x] setup at https://cloud.docker.com/u/spack/repository/docker/spack/packages.spack.io/builds/edit

Follow-up to #11665 and #11621

### CODE REVIEW

File: share/spack/docker/package-index/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+RUN apt-get -yqq update   \
+ && apt-get -yqq install  \
+        bash jq python    \
+ && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /build
+# single, large index file
+RUN ${SPACK_ROOT}/bin/spack list --format version_json > packages.json
+# individual packages split into a tree of :firstLetter/:packageName.json
+COPY share/spack/packages/split.sh ./
```
Comment (alalazo) [inline share/spack/docker/package-index/Dockerfile]: I'm trying this PR and Docker says this file is missing:
```console
Step 11/17 : RUN ${SPACK_ROOT}/bin/spack list --format version_json > packages.json
 ---> Using cache
 ---> ac397f05ce75
Step 12/17 : COPY share/spack/packages/split.sh ./
COPY failed: stat /var/lib/docker/tmp/docker-builder269930503/share/spack/packages/split.sh: no such file or directory

$ ls share/spack/packages/
Dockerfile
```
Looking at the PR it seems it has been moved :thinking:

Comment (alalazo) [inline share/spack/docker/package-index/Dockerfile]:
```suggestion
COPY share/spack/docker/package-index/split.sh ./
```

File: share/spack/docker/package-index/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+RUN ./split.sh
+
+# nginx web service
+FROM nginx:mainline-alpine
+COPY --from=build-env --chown=nginx:nginx /build/packages /build/packages.json /usr/share/nginx/html/api/
+COPY share/spack/packages/cors-header.conf /etc/nginx/conf.d/
```
Comment (alalazo) [inline share/spack/docker/package-index/Dockerfile]:
```suggestion
COPY share/spack/docker/package-index/cors-header.conf /etc/nginx/conf.d/
```

File: share/spack/packages/Dockerfile [hunk 3]
```diff
@@ -1,12 +1,32 @@
+# prepare the package index in form of JSON files
 FROM ubuntu:18.04 AS build-env
+MAINTAINER Spack Maintainers <maintainers@spack.io>
```
Comment (alalazo) [inline share/spack/packages/Dockerfile]: Is this file supposed to be deleted?

File: share/spack/docker/package-index/Dockerfile (hunk 1 above)
Comment (ax3l) [reply share/spack/docker/package-index/Dockerfile]: Yes, please use the `-f` and `.` parameters while sitting in the spack root. learned this as well today.

Comment (ax3l) [reply share/spack/docker/package-index/Dockerfile]: waaaiit... I forgot to commit my last changes. thanks for catching!

File: share/spack/packages/Dockerfile (hunk 3 above)
Comment (ax3l) [reply share/spack/packages/Dockerfile]: oh yes :)

**Comment by ax3l** [comment]
@alalazo thank you for the review, my last updates did not reach GitHub it seams. Sorry for that.
I also added a README now in the same directory, pointing to the related services this interfaces with.

**Comment by alalazo** [comment]
> I am sorry for that.

No need to be, happy to review your PRs :nerd_face:

File: share/spack/docker/package-index/README.rst [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+On each merge to ``develop``, DockerHub builds a new image ``spack/packages.spack.io`` which is configured in:
+  https://cloud.docker.com/u/spack/repository/docker/spack/packages.spack.io/builds/edit
+
+------------
+The REST API
+------------
+
+The API is a simple, file-based JSON index.
+A specific package can be queried via the URI syntax:
+``https://packages.spack.io/api/:firstLetter/:packageName.json``
```
Comment (opadron) [inline share/spack/docker/package-index/README.rst]: This is more curiosity on my part than anything else, but why the `:firstLetter` part?  Is that part of some standard that this service is trying to adhere to?

Comment (ax3l) [reply share/spack/docker/package-index/README.rst]: This is just me trying to implement a very naïve load balancing to split the cost of an 5k inode `ls`.
```
ls var/spack/repos/builtin/packages/ | wc -l
3695
```

Comment (ax3l) [reply share/spack/docker/package-index/README.rst]: I am anticipating 100k+ packages in Spack

File: share/spack/docker/package-index/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ && apt-get -yqq install  \
+        bash jq python    \
+ && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /build
+# single, large index file
+RUN $SPACK_ROOT/bin/spack list --format version_json > packages.json
+# individual packages split into a tree of :firstLetter/:packageName.json
+COPY share/spack/docker/package-index/split.sh ./
+RUN ./split.sh
```
Comment (opadron) [inline share/spack/docker/package-index/Dockerfile]:
```diff
--- COPY share/spack/docker/package-index/split.sh ./
--- RUN ./split.sh
+++ RUN $SPACK_ROOT/share/spack/docker/package-index/split.sh
```

Comment (ax3l) [reply share/spack/docker/package-index/Dockerfile]: Really, that works? It's an external script o.0

Comment (ax3l) [reply share/spack/docker/package-index/Dockerfile]: ah no, I copy all things in xD

**Pull request merged**